src/open_inwoner/accounts/models.py

```diff
@@ -246,30 +244,44 @@ class Meta:
                 name="unique_email_when_not_digid_or_eherkenning",
             ),
-            # UniqueConstraint(
-            #     fields=["bsn"],
-            #     condition=Q(login_type=LoginTypeChoices.digid)
-            #     # not quite sure if we want this (bsn shouldn't be null AND blank)
-            #     & ~(Q(bsn="") | Q(bsn__isnull=True)),
-            #     name="unique_bsn_when_digid",
-            # ),
-            # UniqueConstraint(
-            #     fields=["rsin"],
-            #     condition=Q(login_type=LoginTypeChoices.eherkenning)
-            #     # not quite sure if we want this (rsin shouldn't be null AND blank)
-            #     & ~(Q(rsin="") | Q(rsin__isnull=True)),
-            #     name="unique_rsin_when_eherkenning",
-            # ),
-            # UniqueConstraint(
-            #     fields=["oidc_id"],
-            #     condition=Q(login_type=LoginTypeChoices.oidc),
-            #     name="unique_bsn_when_digid",
-            # ),
+            UniqueConstraint(
+                fields=["bsn"],
+                condition=Q(login_type=LoginTypeChoices.digid),
+                name="unique_bsn_when_login_digid",
+            ),
+            UniqueConstraint(
+                fields=["rsin"],
+                condition=Q(login_type=LoginTypeChoices.eherkenning) & ~Q(rsin=""),
+                name="unique_rsin_when_login_eherkenning",
+            ),
+            UniqueConstraint(
+                fields=["kvk"],
+                condition=Q(login_type=LoginTypeChoices.eherkenning) & ~Q(kvk=""),
+                name="unique_kvk_when_login_eherkenning",
+            ),
+            UniqueConstraint(
+                fields=["oidc_id"],
+                condition=Q(login_type=LoginTypeChoices.oidc),
+                name="unique_oidc_id_when_login_oidc",
+            ),
+            # --- --- ---
+            CheckConstraint(
+                check=Q(bsn="") | Q(login_type=LoginTypeChoices.digid),
+                name="check_bsn_only_set_when_login_digid",
+            ),
+            CheckConstraint(
+                check=Q(oidc_id="") | Q(login_type=LoginTypeChoices.oidc),
+                name="check_oidc_id_only_set_when_login_oidc",
+            ),
+            CheckConstraint(
+                check=(Q(kvk="") & Q(rsin=""))
+                | Q(login_type=LoginTypeChoices.eherkenning),
+                name="check_kvk_or_rsin_only_set_when_login_eherkenning",
+            ),
             # CheckConstraint(
-            #     # maybe this is not correct?
-            #     check=(Q(bsn="") | Q(bsn__isnull=True))
-            #     | ~Q(login_type=LoginTypeChoices.digid),
-            #     name="check_digid_bsn_required_when_digid",
+            #     check=((Q(kvk="") & ~Q(rsin="")) | (~Q(kvk="") & Q(rsin="")))
+            #     | ~Q(login_type=LoginTypeChoices.eherkenning),
+            #     name="check_kvk_or_rsin_exclusive_when_login_eherkenning",
             # ),
```

Comment:
I left this one out, because our current implementation uses both KVK and RSIN (OpenZaak requires RSIN and eSuite KVK)

Reply:
This rule enforces that when a user is using eherkenning you can't have both kvk and rsin set on the same instance.

This was an assumption for review at the time I made this, I'm not 100% into eherkenning so maybe it is legal to have both at same time?

Reply:
It is legal to have both: https://www.kvk.nl/over-het-handelsregister/rsin-nummer/

Apparently all companies (except eenmanszaken) have both KVK nummer and RSIN. I'll create a separate issue to check how we handle the eenmanszaak situation if we connect to ZGW apis right now, because I'm fairly sure that won't work with our current implementation

Reply:
Cool, thanks for checking. Then we could remove the commented-out constraint and replace it with a text comment that reminds the next reader we explicitly didn't constrained this situation because it is legal to have both.

Reply:
On second thought, this eenmanszaak edge case might mean that we cannot add the constraint that users can only have a BSN when they log in using DigiD, because we'll (probably?) have to use the BSN of the owner of the eenmanszaak to fetch Zaken etc.